rust/datafusion/src/physical_plan/planner.rs

```diff
@@ -344,13 +345,33 @@ impl DefaultPhysicalPlanner {
                     JoinType::Left => hash_utils::JoinType::Left,
                     JoinType::Right => hash_utils::JoinType::Right,
                 };
-
-                Ok(Arc::new(HashJoinExec::try_new(
-                    left,
-                    right,
-                    &keys,
-                    &physical_join_type,
-                )?))
+                if ctx_state.config.concurrency > 1 {
+                    let left_expr = keys.iter().map(|x| col(&x.0)).collect();
+                    let right_expr = keys.iter().map(|x| col(&x.1)).collect();
+
+                    // Use hash partition by defualt to parallelize hash joins
+                    Ok(Arc::new(HashJoinExec::try_new(
+                        Arc::new(RepartitionExec::try_new(
```

Comment:
One thing I worry about is that `RepartitionExec` has the potential to (effectively) buffer up the entire result set of the producer makes data faster than the consumer can read it -- e.g. https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/rust/datafusion/src/physical_plan/repartition.rs#L121-L126 it uses unbounded streams.

Thus this particular structure might end up having a tradeoff of memory usage for keeping more cores busy -- which might or might not be the right trade off depending on the system). 

I wonder if you would be amenable to adding a new setting to `ExecutionConfig` that controlled this optimization? 

So like 
```
                if ctx_state.config.concurrency > 1 && config_ctx.config.repartition_joins {
```

Reply:
Yeah I think I saw that happening I think in some cases when testing with in-memory data.
We have to have a solution for that at some point to buffer only up to x batches or x MB something like that.
I am willing to add a config for this although we might also try to reduce the n.r. of knobs we add?

Reply:
> we might also try to reduce the n.r. of knobs we add?

Yeah, lots of knobs isn't ideal... but sometimes they are helpful as an escape hatch / workaround if someone hits an error.

> We have to have a solution for that at some point to buffer only up to x batches or x MB something like that.

I think the challenge when using hash partitioning, in the pathological case you actually *do* have to buffer all the input 

For example,if one input to the hash was all NULLs and the other side had 10M non-null values and then a NULL. You effectively will end up with all the NULLs in a single partition that can't make any progress until the other side has been completely drained. 

The best you can do I think is to make sure your repartitioning algorithm responds to backpressure but also is prepared to buffer / spill the entire thing to disk

Reply:
You are totally right about this.

For Ballista @andygrove was planning to implement shuffle hash join which should be pretty similar to this except that the data/partitions should be shuffled to different workers. When the data is balanced, here adding more nodes should help process more data, even without spilling to disk - given we don't have other sources of high memory usage / unlimited buffering.
A sort merge join implementation would also make sense for Ballista in the longer term and probably would be great to have in DataFusion too (next to any other scalable algorithms).

Reply:
I also added the config setting now btw

Reply:
> A sort merge join implementation would also make sense

fyi: [merge_sort.rs](https://github.com/jorgecarleitao/arrow2/blob/main/src/compute/merge_sort/mod.rs)

Reply:
Thanks @Dandandan  -- will re-check this PR later today